hnn_core/opt_toy_example/general.py

```diff
@@ -0,0 +1,370 @@
+    """Uses gp_minimize optimizer.
+
+       Parameters
+       ----------
+       net : Network
+       params : dictionary
+           Contains parameter names, initial parameters, and constraints.
+       obj_fun : func
+           The objective function.
+        target_statistic : ndarray
+            The recorded dipole.
+        max_iter : int
+            Max number of calls.
+
+       Returns
+       -------
+       opt_params : list
+           Final parameters.
+       obj : list
+           Objective values.
+       net_ : Network
+           Optimized network object.
+    """
+
+    def _obj_func(predicted_params):
+        return obj_fun(net,
+                       params['names'],
+                       target_statistic,
+                       predicted_params,
+                       f_bands,
+                       weights,
+                       _set_params)
+
+    opt_results = gp_minimize(func=_obj_func,
+                              dimensions=params['constraints'],
+                              acq_func='EI',
+                              n_calls=max_iter,
+                              x0=params['initial'],
+                              random_state=64)
+    opt_params = opt_results.x
+    obj = [np.min(opt_results.func_vals[:i]) for i in range(1, max_iter + 1)]
+    # get optimized net
+    net_ = _set_params(net, params['names'], opt_params)
+    return opt_params, obj, net_
+
+
+def _run_opt_cobyla(net, params, obj_fun, target_statistic, max_iter, f_bands,
+                    weights):
+    """Uses fmin_cobyla optimizer.
+
+       Parameters
+       ----------
+       net : Network
+       params : dictionary
+           Contains parameter names, initial parameters, and constraints.
+       obj_fun : func
+           The objective function.
+        target_statistic : ndarray
+            The recorded dipole.
+        max_iter : int
+            Max number of calls.
+
+       Returns
+       -------
+       opt_params : list
+           Final parameters.
+       obj : list
+           Objective values.
+       net_ : Network
+           Optimized network object.
+    """
+
+    def _obj_func(predicted_params):
+        return obj_fun(net,
+                       params['names'],
+                       target_statistic,
+                       predicted_params,
+                       f_bands,
+                       weights,
+                       _set_params)
+
+    opt_results = fmin_cobyla(_obj_func,
+                              cons=params['constraints'],
+                              rhobeg=0.1,
+                              rhoend=1e-4,
+                              x0=params['initial'],
+                              maxfun=max_iter,
+                              catol=0.0)
+    opt_params = opt_results
+    obj = list()
+    # get optimized net
+    net_ = _set_params(net, params['names'], opt_params)
+    return opt_params, obj, net_
+
+
+def _set_params(net, param_names, predicted_params):
```

Comment:
For the generic `set_params()` function that a user passes in, you could have one input called `params_dict` which looks like
`{'mu': 50, 'sigma': 4.5, 'ampa_weight': 2, ...}`

Reply:
and then "hard-code" parameter updated as `net.add_evoked(mu=params_dict['mu'], ...)`